### PR TITLE
Go debugging support

### PR DESCRIPTION
**Pull request by hunterwerlla**
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: jetbrains-ultimate/src/software/aws/toolkits/jetbrains/services/lambda/go/GoSamDebugSupport.kt [hunk 1]
```diff
@@ -3,10 +3,31 @@
...
+import software.aws.toolkits.core.utils.getLogger
+import software.aws.toolkits.core.utils.warn
+import software.aws.toolkits.jetbrains.core.utils.buildList
 import software.aws.toolkits.jetbrains.services.lambda.execution.sam.SamDebugSupport
 import software.aws.toolkits.jetbrains.services.lambda.execution.sam.SamRunningState
+import software.aws.toolkits.jetbrains.utils.ApplicationThreadPoolScope
+import java.net.InetSocketAddress
+import java.nio.file.Files
 
 class GoSamDebugSupport : SamDebugSupport {
```
Comment (abrooksv) [inline jetbrains-ultimate/src/software/aws/toolkits/jetbrains/services/lambda/go/GoSamDebugSupport.kt:32]: Should we override isSupported to be false (and update it to a min sam version once its patched?)

Comment (hunterwerlla) [reply jetbrains-ultimate/src/software/aws/toolkits/jetbrains/services/lambda/go/GoSamDebugSupport.kt:32]: I was going to wait for it to be fixed to merge go support, we can decide what we want to do when we get closer to release though

File: jetbrains-ultimate/it/software/aws/toolkits/jetbrains/services/lambda/go/GoLocalRunConfigurationIntegrationTest.kt [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        runInEdtAndWait {
+            fixture.openFileInEditor(psiFile.virtualFile)
+        }
+
+        credentialManager.addCredentials(mockId, mockCreds)
+        WebStormTestUtils.ensureBuiltInServerStarted()
+    }
+
+    @Test
+    fun samIsExecuted() {
```
Comment (abrooksv) [inline jetbrains-ultimate/it/software/aws/toolkits/jetbrains/services/lambda/go/GoLocalRunConfigurationIntegrationTest.kt:87]: Can we add the env vars tests that we have in python?

File: jetbrains-ultimate/it/software/aws/toolkits/jetbrains/services/lambda/go/GoLocalRunConfigurationIntegrationTest.kt [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        PsiTestUtil.addModule(projectRule.project, ModuleType.EMPTY, "main", fixture.tempDirFixture.findOrCreateDir("."))
+
+        val psiFile = fixture.addFileToProject("hello-world/main.go", fileContents)
+
+        runInEdtAndWait {
+            fixture.openFileInEditor(psiFile.virtualFile)
+        }
+
+        credentialManager.addCredentials(mockId, mockCreds)
+        WebStormTestUtils.ensureBuiltInServerStarted()
```
Comment (abrooksv) [inline jetbrains-ultimate/it/software/aws/toolkits/jetbrains/services/lambda/go/GoLocalRunConfigurationIntegrationTest.kt:83]: Why is this in here?

File: jetbrains-ultimate/src/software/aws/toolkits/jetbrains/services/lambda/go/GoSamDebugSupport.kt [hunk 4]
```diff
@@ -15,6 +36,92 @@ class GoSamDebugSupport : SamDebugSupport {
...
+            add("-delveAPI=2")
+        }
+    }
+
+    private fun copyDlv(): String {
+        val dlvFolder = getDelve()
+        val directory = Files.createTempDirectory("goDebugger")
+        Files.copy(dlvFolder.resolve("dlv").toPath(), directory.resolve("dlv"))
+        // Delve that comes packaged with the IDE does not have the executable flag set
+        directory.resolve("dlv").toFile().setExecutable(true)
```
Comment (abrooksv) [inline jetbrains-ultimate/src/software/aws/toolkits/jetbrains/services/lambda/go/GoSamDebugSupport.kt:120]: Hmm.... What happens if host OS is Windows?

Comment (hunterwerlla) [reply jetbrains-ultimate/src/software/aws/toolkits/jetbrains/services/lambda/go/GoSamDebugSupport.kt:120]: Actually this works on windows as is, SAM cli must be doing something funky on windows but it works:
Capture

**Review by abrooksv — APPROVED** [review]

**Pull request merged**